ir/src/passes/inlining.rs

```diff
@@ -0,0 +1,211 @@
+use std::{collections::BTreeMap, ops::ControlFlow};
+
+use air_pass::Pass;
+//use miden_diagnostics::DiagnosticsHandler;
+
+use crate::{MirGraph, NodeIndex, Operation};
+
+//pub struct Inlining<'a> {
+//     #[allow(unused)]
+//     diagnostics: &'a DiagnosticsHandler,
+//}
+
+pub struct Inlining {}
```

Comment:
For this pass, you can again use a post-order visit of the graph, inlining the regions of structured ops (e.g. comprehensions, functions) into the containing region, by replacing all uses of entry block arguments with the operands used at the call site.

The intrusive linked lists containing operations in a `Block` can be spliced/split and stolen relative to a cursor, so it is quite efficient to do the actual mechanical inlining.

It might not also be all that important to do inlining with the representation I outlined above, since I suspect it is relatively straightforward to lower to AIR without it, but some optimizations will be region or block-local, so that's the only tradeoff.

Reply:
> The intrusive linked lists containing operations in a Block can be spliced/split and stolen relative to a cursor, so it is quite efficient to do the actual mechanical inlining.

Wouldn't reusing existing nodes by splicing them during inlining cause conflicts in case a function is called twice (or more) with different arguments?

The current implementation duplicates those nodes for this reason. The main problem in doing that is maintaining the unique property globally, which should be fixed once a regionalized uniquing strategy is implemented.

Reply:
> Wouldn't reusing existing nodes by splicing them during inlining cause conflicts in case a function is called twice (or more) with different arguments?

Sorry, I didn't mean stealing then splicing in this context, but rather that splicing the _cloned_ list of ops into the middle of the op list you want to inline into, is very efficient due to the intrusive lists, only requiring writes to two pointers.

The exception would be things like a comprehension that turns out to only have a single iteration, you wouldn't need to clone in that case, but could just steal the list from the comprehension body.